File: mobile/src/main/java/com/priyanka/brightnesscontroller/MainActivity.java
package com.priyanka.brightnesscontroller;

import android.Manifest;
import android.annotation.SuppressLint;
import android.annotation.TargetApi;
import android.content.ActivityNotFoundException;
import android.content.ContentResolver;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.provider.Settings;
import android.provider.Settings.System;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;
import android.widget.Button;
import android.widget.SeekBar;
import android.widget.Toast;

public class MainActivity extends AppCompatActivity {

    //Write setting Permission
    private static final int WRITE_SETTINGS_REQUEST_CODE = 121;
    @SuppressLint("StaticFieldLeak")
    private static Context context;
    //a variable to store the system brightness
    private int brightness;
    //the content resolver used as a handle to the system's settings
    private ContentResolver cResolver;
    //a window object, that will store a reference to the current window
    private Window window;

    @TargetApi(Build.VERSION_CODES.M)
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        context = MainActivity.this;
        //checking write setting permission
        try {
            showRequestPermissionWriteSettings();
        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            Log.i("Final Block Executed", "FINAL BLOCK EXECUTED");
        }

    }

    @Override
    protected void onResume() {
        super.onResume();
        try {
            setContentView(R.layout.activity_main);


            Button buttonRate = (Button) findViewById(R.id.btn_rate);
            Button buttonShare = (Button) findViewById(R.id.btn_share);
            Button buttonFeedBack = (Button) findViewById(R.id.btn_feedback);


            //get the seek bar from main.xml file
            SeekBar brightBar = (SeekBar) findViewById(R.id.seek_brightbar);


            //get the content resolver
            cResolver = getContentResolver();

            //get the current window
            window = getWindow();

            //seek bar settings//
            //sets the range between 0 and 255
            brightBar.setMax(255);
            //set the seek bar progress to 1
            brightBar.setKeyProgressIncrement(1);
            try {
                //get the current system brightness
                brightness = Settings.System.getInt(cResolver, Settings.System.SCREEN_BRIGHTNESS);

                //set the system brightness value, and if active, disable auto mode.
                Settings.System.putInt(cResolver, Settings.System.SCREEN_BRIGHTNESS, brightness);
                //Settings.System.putInt(cResolver, Settings.System.SCREEN_BRIGHTNESS_MODE, Settings.System.SCREEN_BRIGHTNESS_MODE_MANUAL);

            } catch (Settings.SettingNotFoundException e) {
                //throw an error case it couldn't be retrieved
                Log.e("Error", "Cannot access system brightness");
                e.printStackTrace();
            } catch (Exception e) {
                e.printStackTrace();
            }

            //sets the progress of the seek bar based on the system's brightness
            brightBar.setProgress(brightness);


            //register OnSeekBarChangeListener, so it can actually change values
            brightBar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
                @Override
                public void onStopTrackingTouch(SeekBar seekBar) {
                    //set the system brightness using the brightness variable value
                    System.putInt(cResolver, Settings.System.SCREEN_BRIGHTNESS, brightness);

                    //preview brightness changes at this window
                    //get the current window attributes
                    WindowManager.LayoutParams layoutpars = window.getAttributes();
                    //set the brightness of this window
                    layoutpars.screenBrightness = brightness / (float) 255;

                    //apply attribute changes to this window
                    window.setAttributes(layoutpars);
                    window.setDimAmount((float) 0.0);
                    window.setFlags(WindowManager.LayoutParams.FLAG_DIM_BEHIND,
                            WindowManager.LayoutParams.FLAG_DIM_BEHIND);
                }

                @Override
                public void onStartTrackingTouch(SeekBar seekBar) {
                }

                @Override
                public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                    //sets the minimal brightness level
                    //if seek bar is 20 or any value below
                    if (progress <= 20) {
                        //set the brightness to 20
                        brightness = 0;
                    } else //brightness is greater than 20
                    {
                        //sets brightness variable based on the progress bar
                        brightness = progress;
                    }
                }
            });


            //-------------------------------------------------functionality for buttons
            buttonRate.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    Uri uri = Uri.parse("market://details?id=" + context.getPackageName());
                    Intent goToMarket = new Intent(Intent.ACTION_VIEW, uri);
                    // To count with Play market backstack, After pressing back button,
                    // to taken back to our application, we need to add following flags to intent.
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                        goToMarket.addFlags(Intent.FLAG_ACTIVITY_NO_HISTORY |
                                Intent.FLAG_ACTIVITY_NEW_DOCUMENT |
                                Intent.FLAG_ACTIVITY_MULTIPLE_TASK);
                    }
                    try {
                        startActivity(goToMarket);
                    } catch (ActivityNotFoundException e) {
                        startActivity(new Intent(Intent.ACTION_VIEW,
                                Uri.parse("http://play.google.com/store/apps/details?id=" + context.getPackageName())));
                    } catch (Exception e) {
                        e.printStackTrace();
                    } finally {
                        Log.i("Final Block Executed", "FINAL BLOCK EXECUTED");
                    }
                }
            });
            buttonFeedBack.setOnClickListener(new View.OnClickListener() {
                @SuppressLint("LongLogTag")
                @Override
                public void onClick(View v) {
                    String email = "appfeedbackpriyanka@gmail.com";
                    String Subject = "User Feedback : Brightness Controller Application";
                    Intent emailIntent = new Intent(Intent.ACTION_SEND);
                    emailIntent.setData(Uri.parse("mailto:"));
                    emailIntent.setType("text/plain");
                    emailIntent.putExtra(Intent.EXTRA_EMAIL, new String[]{email});
                    emailIntent.putExtra(Intent.EXTRA_SUBJECT, Subject);
                    emailIntent.putExtra(Intent.EXTRA_TEXT, "");
                    try {
                        startActivity(Intent.createChooser(emailIntent, "Send mail"));
                        finish();
                        Log.i("Finished sending email...", "");
                    } catch (android.content.ActivityNotFoundException ex) {
                        Toast.makeText(MainActivity.this, "There is no email client installed.", Toast.LENGTH_SHORT).show();
                    } catch (Exception e) {
                        Toast.makeText(MainActivity.this, "There is no email client installed.", Toast.LENGTH_SHORT).show();
                    } finally {
                        Log.i("Final Block Executed", "FINAL BLOCK EXECUTED");
                    }
                }
            });

            buttonShare.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    String shareBody = "https://play.google.com/store/apps/details?id=" + context.getPackageName();
                    Intent sharingIntent = new Intent(android.content.Intent.ACTION_SEND);
                    sharingIntent.setType("text/plain");
                    sharingIntent.putExtra(android.content.Intent.EXTRA_SUBJECT, context.getResources().getString(R.string.app_name) +
                            "\n\n(Open it in Google Play Store to Download the Application)" + context.getResources().getString(R.string.app_name));
                    sharingIntent.putExtra(android.content.Intent.EXTRA_TEXT,
                            context.getResources().getString(R.string.app_name) + ":\n\n(Open it in Google Play Store to Download the Application)\n"
                                    + shareBody);
                    try {
                        startActivity(Intent.createChooser(sharingIntent, "Share via"));
                    } catch (Exception e) {
                        Toast.makeText(MainActivity.this, "There is no app installed for sharing this application.", Toast.LENGTH_SHORT).show();
                    } finally {
                        Log.i("Final Block Executed", "FINAL BLOCK EXECUTED");
                    }

                }
            });

        }//---------------------try close
        catch (Exception e) {
            e.printStackTrace();
        } finally {
            Log.i("Final Block Executed", "FINAL BLOCK EXECUTED");
        }

    }

    @Override
    public void onBackPressed() {
        super.onBackPressed();
        finish();
    }

    //system settings permission
    @TargetApi(Build.VERSION_CODES.M)
    private void showRequestPermissionWriteSettings() {
        try {
            // for Settings.ACTION_MANAGE_WRITE_SETTINGS: Settings.System.canWrite
            // CommonsWare's blog post:https://commonsware.com/blog/2015/08/17/random-musings-android-6p0-sdk.html
            boolean hasSelfPermission;
            //boolean hasSelfPermission = Settings.System.canWrite(this);
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                hasSelfPermission = Settings.System.canWrite(this);
            } else {
                hasSelfPermission = ContextCompat.checkSelfPermission(context, Manifest.permission.WRITE_SETTINGS) == PackageManager.PERMISSION_GRANTED;
            }

            if (hasSelfPermission) {
         /*  new AlertDialog.Builder(this)
                    .setCancelable(false)
                    .setTitle("Permission Brightness Write Settings")
                    .setMessage("\r\n" + "granted")
                    .setPositiveButton(android.R.string.ok, null)
                    .show();*/
            } else {
                Intent intent = new Intent(Settings.ACTION_MANAGE_WRITE_SETTINGS,
                        Uri.parse("package:" + getPackageName()));
                startActivityForResult(intent, WRITE_SETTINGS_REQUEST_CODE);
            }
        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            Log.i("Final Block Executed", "FINAL BLOCK EXECUTED");
        }

    }


    @TargetApi(Build.VERSION_CODES.M)
    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        try {
            if (requestCode == WRITE_SETTINGS_REQUEST_CODE) {

                final boolean hasSelfPermission = Settings.System.canWrite(this);
                new AlertDialog.Builder(this)
                        .setCancelable(false)
                        .setTitle(context.getResources().getString(R.string.text_brightness_access))
                  /*  .setMessage((hasSelfPermission ? "granted" : "not granted") + "\r\n\r\n" +
                        context.getResources().getString(R.string.text_request))*/
                        .setMessage((hasSelfPermission ? "Granted" : "Not Granted") + "\r\n\r\n" +
                                context.getResources().getString(R.string.text_request))
                        .setPositiveButton(android.R.string.ok, new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialog, int which) {
                                if (!hasSelfPermission) {
                                    //onBackPressed();
                                    showRequestPermissionWriteSettings();
                                }
                            }
                        })
                        .show();
            }
        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            Log.i("Final Block Executed", "FINAL BLOCK EXECUTED");
        }

    }
}

File: mobile/src/main/java/com/priyanka/brightnesscontroller/MainActivity.kt
package com.priyanka.brightnesscontroller

import android.content.ActivityNotFoundException
import android.content.Intent
import android.os.Bundle
import android.provider.Settings
import android.widget.Toast
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.activity.enableEdgeToEdge
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.app.AppCompatDelegate
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.safeDrawingPadding
import androidx.compose.material3.Button
import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.Slider
import androidx.compose.material3.SliderDefaults
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableFloatStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.unit.dp
import androidx.core.net.toUri
import androidx.core.os.LocaleListCompat
import com.priyanka.brightnesscontroller.ui.theme.BrightnessControllerTheme
import com.priyanka.brightnesscontroller.ui.theme.TealPrimary
import com.priyanka.brightnesscontroller.ui.theme.TealPrimaryDark
import org.koin.android.ext.android.inject

class MainActivity : AppCompatActivity() {

    private val brightnessController: BrightnessController by inject()

    private val permissionLauncher = registerForActivityResult(
        ActivityResultContracts.StartActivityForResult()
    ) {
        if (!brightnessController.hasWriteSettingsPermission()) {
            Toast.makeText(this, getString(R.string.permission_denied), Toast.LENGTH_LONG).show()
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        enableEdgeToEdge()
        super.onCreate(savedInstanceState)
        
        // Request permissions on startup if not granted (Required for API 23+)
        if (!brightnessController.hasWriteSettingsPermission()) {
            val intent = Intent(Settings.ACTION_MANAGE_WRITE_SETTINGS, "package:$packageName".toUri())
            permissionLauncher.launch(intent)
        }

        setContent {
            BrightnessControllerTheme {
                Surface(
                    modifier = Modifier
                        .fillMaxSize()
                        .safeDrawingPadding(),
                    color = MaterialTheme.colorScheme.background
                ) {
                    BrightnessScreen(brightnessController)
                }
            }
        }
    }
}

data class Language(val name: String, val code: String)

val supportedLanguages = listOf(
    Language("Default (System)", ""),
    Language("English", "en"),
    Language("Español (Spanish)", "es"),
    Language("Français (French)", "fr"),
    Language("Indonesian", "id"),
    Language("اردو (Urdu)", "ur"),
    Language("عربي (Arabic)", "ar"),
    Language("ไทย (Thai)", "th"),
    Language("हिन्दी (Hindi)", "hi"),
    Language("ଓଡ଼ିଆ (Odia)", "or"),
    Language("తెలుగు (Telugu)", "te"),
    Language("ಕನ್ನಡ (Kannada)", "kn"),
    Language("मराठी (Marathi)", "mr"),
    Language("ತುಳು (Tulu)", "tcy"),
    Language("বাংলা (Bengali)", "bn")
)

@Composable
fun LanguageSelectionDropdown() {
    var expanded by remember { mutableStateOf(false) }
    val currentLocale = AppCompatDelegate.getApplicationLocales().toLanguageTags()
    val selectedLanguage = supportedLanguages.find { it.code == currentLocale } ?: supportedLanguages[0]

    Column(
        horizontalAlignment = Alignment.CenterHorizontally,
        modifier = Modifier.padding(bottom = 16.dp)
    ) {
        Text(
            text = stringResource(R.string.select_language),
            style = MaterialTheme.typography.titleMedium,
            modifier = Modifier.padding(bottom = 8.dp)
        )
        Box {
            OutlinedButton(
                onClick = { expanded = true },
                modifier = Modifier.height(56.dp) // Taller button for easier tapping
            ) {
                Text(
                    text = "🌐 ${selectedLanguage.name}",
                    style = MaterialTheme.typography.bodyLarge
                )
            }
            DropdownMenu(
                expanded = expanded,
                onDismissRequest = { expanded = false }
            ) {
                supportedLanguages.forEach { language ->
                    DropdownMenuItem(
                        text = { Text(language.name) },
                        onClick = {
                            val appLocale: LocaleListCompat = if (language.code.isEmpty()) {
                                LocaleListCompat.getEmptyLocaleList()
                            } else {
                                LocaleListCompat.forLanguageTags(language.code)
                            }
                            AppCompatDelegate.setApplicationLocales(appLocale)
                            expanded = false
                        }
                    )
                }
            }
        }
    }
}

@Composable
fun BrightnessScreen(brightnessController: BrightnessController) {
    val context = LocalContext.current
    var brightness by remember { mutableFloatStateOf(brightnessController.getSystemBrightness().toFloat()) }
    
    val shareSubject = stringResource(R.string.share_subject)
    val shareTextTemplate = stringResource(R.string.share_text)
    val noShareApp = stringResource(R.string.no_share_app)
    val emailSubject = stringResource(R.string.email_subject)
    val noEmailApp = stringResource(R.string.no_email_app)

    Column(
        modifier = Modifier
            .fillMaxSize()
            .padding(16.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.Center
    ) {
        LanguageSelectionDropdown()

        Text(
            text = stringResource(R.string.app_name),
            style = MaterialTheme.typography.headlineMedium
        )
        
        Spacer(modifier = Modifier.height(32.dp))

        Text(text = stringResource(R.string.brightness_level, brightness.toInt()))

        Slider(
            value = brightness,
            onValueChange = { newValue ->
                brightness = newValue
                if (brightnessController.hasWriteSettingsPermission()) {
                    brightnessController.setSystemBrightness(newValue.toInt())
                }

                var activity = context as? ComponentActivity
                var ctx = context
                while (activity == null && ctx is android.content.ContextWrapper) {
                    ctx = ctx.baseContext
                    activity = ctx as? ComponentActivity
                }
                if (activity?.window != null) {
                    val attrs = activity.window.attributes
                    attrs.screenBrightness = newValue / 255f
                    activity.window.attributes = attrs
                }
            },
            valueRange = 0f..255f,
            modifier = Modifier
                .fillMaxWidth()
                .padding(horizontal = 16.dp),
            // FORCE COLORS TO RESET THE THUMB SHAPE
            colors = SliderDefaults.colors(
                thumbColor = TealPrimaryDark,           // This forces the thumb to be your Teal
                activeTrackColor = TealPrimary,      // The line to the left
                inactiveTrackColor = TealPrimary.copy(alpha = 0.30f), // The line to the right
                activeTickColor = Color.Transparent, // REMOVES THE DOTS
                inactiveTickColor = Color.Transparent // REMOVES THE DOTS
            )
        )
        Spacer(modifier = Modifier.height(48.dp))

        Button(onClick = {
            val uri = "market://details?id=${context.packageName}".toUri()
            val goToMarket = Intent(Intent.ACTION_VIEW, uri).apply {
                addFlags(Intent.FLAG_ACTIVITY_NO_HISTORY or Intent.FLAG_ACTIVITY_NEW_DOCUMENT or Intent.FLAG_ACTIVITY_MULTIPLE_TASK)
            }
            try {
                context.startActivity(goToMarket)
            } catch (_: ActivityNotFoundException) {
                context.startActivity(Intent(Intent.ACTION_VIEW, "http://play.google.com/store/apps/details?id=${context.packageName}".toUri()))
            }
        }, modifier = Modifier.fillMaxWidth()) {
            Text(text = stringResource(R.string.rate_app))
        }

        Spacer(modifier = Modifier.height(8.dp))

        Button(onClick = {
            val shareBody = "https://play.google.com/store/apps/details?id=${context.packageName}"
            val sharingIntent = Intent(Intent.ACTION_SEND).apply {
                type = "text/plain"
                putExtra(Intent.EXTRA_SUBJECT, shareSubject)
                putExtra(Intent.EXTRA_TEXT, shareTextTemplate.format(shareBody))
            }
            try {
                context.startActivity(Intent.createChooser(sharingIntent, "Share via"))
            } catch (_: Exception) {
                Toast.makeText(context, noShareApp, Toast.LENGTH_SHORT).show()
            }
        }, modifier = Modifier.fillMaxWidth()) {
            Text(text = stringResource(R.string.share_app))
        }

        Spacer(modifier = Modifier.height(8.dp))

        Button(onClick = {
            val emailIntent = Intent(Intent.ACTION_SENDTO).apply {
                data = "mailto:".toUri()
                putExtra(Intent.EXTRA_EMAIL, arrayOf("appfeedbackpriyanka@gmail.com"))
                putExtra(Intent.EXTRA_SUBJECT, emailSubject)
            }
            try {
                context.startActivity(Intent.createChooser(emailIntent, "Send mail"))
            } catch (_: ActivityNotFoundException) {
                Toast.makeText(context, noEmailApp, Toast.LENGTH_SHORT).show()
            }
        }, modifier = Modifier.fillMaxWidth()) {
            Text(text = stringResource(R.string.feedback))
        }
    }
}
